File: copilot-ui/routes/repoAssets.js
'use strict';

const fs = require('fs');
const path = require('path');
const { sendJson, readJsonBody } = require('./_helpers');

/**
 * Repo Assets route module.
 * Discovers agents/skills/config files in a repository and manages
 * installation status per harness via the elegyDB.
 */

const HARNESS_NAMES = ['opencode', 'codex', 'copilot', 'antigravity'];

const DISCOVERY_PATTERNS = [
  // Root config files
  { pattern: 'AGENTS.md', kind: 'config', sourceHarness: null },
  { pattern: 'guidelines.md', kind: 'config', sourceHarness: null },

  // OpenCode
  { glob: '.opencode/agents/*.agent.md', kind: 'agent', sourceHarness: 'opencode' },
  { glob: '.opencode/skills/*/SKILL.md', kind: 'skill', sourceHarness: 'opencode' },

  // Codex
  { glob: '.codex/agents/*.md', kind: 'agent', sourceHarness: 'codex' },
  { glob: '.codex/agents/*.agent.md', kind: 'agent', sourceHarness: 'codex' },
  { glob: '.codex/skills/*/SKILL.md', kind: 'skill', sourceHarness: 'codex' },

  // Copilot
  { glob: '.copilot/agents/*.agent.md', kind: 'agent', sourceHarness: 'copilot' },
  { glob: '.copilot/skills/*/SKILL.md', kind: 'skill', sourceHarness: 'copilot' },

  // Antigravity
  { glob: '.gemini/agents/*.md', kind: 'agent', sourceHarness: 'antigravity' },
  { glob: '.gemini/agents/*.agent.md', kind: 'agent', sourceHarness: 'antigravity' },
  { glob: '.gemini/skills/*/SKILL.md', kind: 'skill', sourceHarness: 'antigravity' },

  // Repo-level (no harness prefix)
  { glob: 'skills/*/SKILL.md', kind: 'skill', sourceHarness: null },
  { glob: 'agents/*.agent.md', kind: 'agent', sourceHarness: null },

  // Generated mirrors (agents) and canonical repo-local (github)
  { glob: '.agents/skills/*/SKILL.md', kind: 'skill', sourceHarness: 'agents' },
  { glob: '.github/skills/*/SKILL.md', kind: 'skill', sourceHarness: 'copilot' },
];

function isNonEmptyString(value) {
  return typeof value === 'string' && value.trim().length > 0;
}

/**
 * Filter dbAssets to only those matching a given assetId.
 * Uses the actual column name `asset_id` (not `assetId`).
 */
function getDbAssetsForAsset(dbAssets, assetId) {
  return (dbAssets || []).filter(
    dbAsset => String(dbAsset.asset_id || '') === String(assetId || '')
  );
}

/**
 * Build the set of harness entries with installation status from elegyDb.
 * The repo_assets table uses per-row design: one row per (repo_path, asset_id, harness).
 */
function buildHarnessStatuses(elegyDb, repoPath, assetId) {
  const statuses = HARNESS_NAMES.map((harness) => ({
    harness,
    installed: false,
    installedAt: null,
  }));

  if (!elegyDb || typeof elegyDb.getRepoAssets !== 'function') {
    return statuses;
  }

  try {
    const dbAssets = elegyDb.getRepoAssets(repoPath);
    if (!Array.isArray(dbAssets)) return statuses;

    const matchingAssets = getDbAssetsForAsset(dbAssets, assetId);

    return HARNESS_NAMES.map((harness) => {
      const matching = matchingAssets.find(
        (dbAsset) => String(dbAsset.harness || '').toLowerCase() === harness.toLowerCase()
      );
      return {
        harness,
        installed: !!matching,
        installedAt: matching ? matching.installed_at || null : null,
      };
    });
  } catch {
    // If db query fails, return default (all not installed)
  }

  return statuses;
}

/**
 * Match a file path against discovery patterns.
 */
function matchPattern(filePath) {
  const normalized = filePath.replace(/\\/g, '/');

  for (const dp of DISCOVERY_PATTERNS) {
    if (dp.pattern) {
      // Exact filename match (root files)
      if (normalized === dp.pattern) {
        return dp;
      }
    }
    if (dp.glob) {
      // Convert simple glob to regex
      const regexStr = dp.glob
        .replace(/\./g, '\\.')
        .replace(/\*/g, '[^/]+');
      const regex = new RegExp('^' + regexStr + '$');
      if (regex.test(normalized)) {
        return dp;
      }
    }
  }

  return null;
}

/**
 * Scan a specific dir for matching files.
 */
function scanDir(repoRoot, results, baseDir) {
  if (!fs.existsSync(baseDir)) return;

  let entries;
  try {
    entries = fs.readdirSync(baseDir, { withFileTypes: true });
  } catch {
    return;
  }

  for (const entry of entries) {
    const fullPath = path.join(baseDir, entry.name);
    const relativePath = path.relative(repoRoot, fullPath).replace(/\\/g, '/');

    if (entry.isDirectory()) {
      if (entry.name === 'node_modules' || entry.name === '.git') continue;
      // Recursively scan subdirs up to reasonable depth
      const depth = relativePath.split('/').length;
      if (depth > 8) continue;
      scanDir(repoRoot, results, fullPath);
    } else if (entry.isFile()) {
      const match = matchPattern(relativePath);
      if (!match) continue;

      try {
        const stat = fs.statSync(fullPath);
        const assetId = relativePath;

        results.push({
          id: assetId,
          name: entry.name,
          kind: match.kind,
          path: relativePath,
          sourceHarness: match.sourceHarness,
          filePath: fullPath,
          size: stat.size,
          modifiedAt: stat.mtime.toISOString(),
        });
      } catch {
        // skip unreadable files
      }
    }
  }
}

/**
 * GET /api/repo-assets/discover?repoPath=...
 */
async function handleRepoAssetsDiscover(ctx, deps) {
  const { res, u, elegyDb } = ctx;
  const { sendJson: json } = deps;
  const repoPath = u.searchParams.get('repoPath');

  if (!isNonEmptyString(repoPath)) {
    json(res, 400, { error: 'repoPath query parameter is required' });
    return;
  }

  const root = repoPath.trim();
  if (!fs.existsSync(root)) {
    json(res, 404, { error: 'Repository path not found' });
    return;
  }

  try {
    const discovered = [];

    // Scan root files
    for (const dp of DISCOVERY_PATTERNS) {
      if (dp.pattern) {
        const fullPath = path.join(root, dp.pattern);
        if (fs.existsSync(fullPath)) {
          try {
            const stat = fs.statSync(fullPath);
            discovered.push({
              id: dp.pattern,
              name: path.basename(dp.pattern),
              kind: dp.kind,
              path: dp.pattern,
              sourceHarness: dp.sourceHarness,
              filePath: fullPath,
              size: stat.size,
              modifiedAt: stat.mtime.toISOString(),
            });
          } catch { /* skip */ }
        }
      }
    }

    // Scan recursive dirs
    scanDir(root, discovered, root);

    // Deduplicate by id (path)
    const seen = new Set();
    const assets = [];
    for (const asset of discovered) {
      if (seen.has(asset.id)) continue;
      seen.add(asset.id);

      assets.push({
        ...asset,
        harnesses: buildHarnessStatuses(elegyDb, root, asset.id),
      });
    }

    // Sort by path
    assets.sort((a, b) => String(a.path || '').localeCompare(String(b.path || '')));

    // Also include worktree info
    let worktrees = [];
    if (elegyDb && typeof elegyDb.listWorktreesByRepo === 'function') {
      try {
        worktrees = elegyDb.listWorktreesByRepo(root) || [];
      } catch (e) { /* ignore */ }
    }

    json(res, 200, {
      repoPath: root,
      assets,
      availableHarnesses: [...HARNESS_NAMES],
      worktrees: worktrees.map(w => ({
        path: w.path,
        branch: w.branch,
        source: w.source,
        status: w.status,
        sessionCount: w.session_count || 0,
        lastActivityAt: w.last_activity_at,
      })),
      count: assets.length,
    });
  } catch (error) {
    json(res, 500, { error: String(error.message || error) });
  }
}

/**
 * POST /api/repo-assets/install
 * Body: { repoPath, assetId, harness }
 */
async function handleRepoAssetsInstall(ctx, deps) {
  const { res, req, elegyDb } = ctx;
  const { sendJson: json } = deps;

  let body;
  try {
    body = await readJsonBody(req);
  } catch (err) {
    const statusCode = err.statusCode || 400;
    json(res, statusCode, { error: String(err.message || 'Invalid request body') });
    return;
  }

  const { repoPath, assetId, harness } = body || {};

  if (!isNonEmptyString(repoPath)) {
    json(res, 400, { error: 'repoPath is required' });
    return;
  }
  if (!isNonEmptyString(assetId)) {
    json(res, 400, { error: 'assetId is required' });
    return;
  }
  if (!isNonEmptyString(harness)) {
    json(res, 400, { error: 'harness is required' });
    return;
  }

  if (!HARNESS_NAMES.includes(harness)) {
    json(res, 400, { error: `Invalid harness. Must be one of: ${HARNESS_NAMES.join(', ')}` });
    return;
  }

  // Verify asset exists in repo
  const root = repoPath.trim();
  const assetPath = path.join(root, assetId);
  if (!fs.existsSync(assetPath)) {
    json(res, 404, { error: 'Asset file not found in repository' });
    return;
  }

  if (!elegyDb || typeof elegyDb.upsertRepoAsset !== 'function') {
    json(res, 503, { error: 'Database not available for recording installation', code: 'elegy_db_not_ready' });
    return;
  }

  try {
    const installedAt = new Date().toISOString();
    const result = elegyDb.upsertRepoAsset({
      repo_path: root,
      asset_id: assetId,
      harness,
      asset_kind: body.assetKind || 'unknown',
      repo_id: body.repoId || null,
      installed_at: installedAt,
      updated_at: installedAt,
      source_path: body.sourcePath || null,
    });

    json(res, 200, {
      ok: true,
      assetId,
      harness,
      installedAt: result ? (result.installed_at || installedAt) : installedAt,
    });
  } catch (error) {
    json(res, 500, { error: String(error.message || error), code: 'elegy_db_write_error' });
  }
}

function register(context = {}) {
  const deps = { sendJson };

  return [
    {
      method: 'GET',
      path: '/api/repo-assets/discover',
      handler: (ctx) => handleRepoAssetsDiscover(ctx, deps),
    },
    {
      method: 'POST',
      path: '/api/repo-assets/install',
      handler: (ctx) => handleRepoAssetsInstall(ctx, deps),
    },
  ];
}

module.exports = { register };

File: native/runtime/src/routes/repo_assets.rs
use axum::{Router, routing::{get, post}, extract::{State, Query}, Json};
use serde::Deserialize;
use std::path::{Path, PathBuf};
use crate::app::AppState;
use crate::error::ApiError;

#[derive(Deserialize)]
struct RepoPathQuery {
    #[serde(rename = "repoPath")]
    repo_path: Option<String>,
}

#[derive(Deserialize)]
struct InstallBody {
    #[serde(rename = "repoPath")]
    repo_path: Option<String>,
    #[serde(rename = "assetId")]
    asset_id: Option<String>,
    #[serde(rename = "sourcePath")]
    source_path: Option<String>,
    harness: Option<String>,
}

// Known harness names
const HARNESS_NAMES: &[&str] = &["opencode", "codex", "copilot", "antigravity"];

pub fn router(state: AppState) -> Router {
    Router::new()
        .route("/api/repo-assets/discover", get(discover))
        .route("/api/repo-assets/install", post(install))
        .with_state(state)
}

fn scan_assets_dir(repo_root: &Path, base_dir: &Path, results: &mut Vec<serde_json::Value>, depth: u32) {
    if depth > 8 || !base_dir.is_dir() {
        return;
    }
    if let Ok(entries) = std::fs::read_dir(base_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            let name = entry.file_name().to_string_lossy().to_string();
            if name == "node_modules" || name == ".git" {
                continue;
            }
            let relative = path.strip_prefix(repo_root)
                .unwrap_or(&path)
                .to_string_lossy()
                .to_string()
                .replace('\\', "/");

            if let Ok(ftype) = entry.file_type() {
                if ftype.is_dir() {
                    scan_assets_dir(repo_root, &path, results, depth + 1);
                } else if ftype.is_file() {
                    // Match known asset patterns
                    let matched = match relative.as_str() {
                        "AGENTS.md" | "guidelines.md" => Some(("config", None)),
                        _ => {
                            // Check glob-style patterns
                            if relative.starts_with(".opencode/agents/") && relative.ends_with(".agent.md") {
                                Some(("agent", Some("opencode")))
                            } else if relative.starts_with(".opencode/skills/") && relative.ends_with("/SKILL.md") {
                                Some(("skill", Some("opencode")))
                            } else if relative.starts_with(".codex/agents/") && (relative.ends_with(".md") || relative.ends_with(".agent.md")) {
                                Some(("agent", Some("codex")))
                            } else if relative.starts_with(".codex/skills/") && relative.ends_with("/SKILL.md") {
                                Some(("skill", Some("codex")))
                            } else if relative.starts_with(".copilot/agents/") && relative.ends_with(".agent.md") {
                                Some(("agent", Some("copilot")))
                            } else if relative.starts_with(".copilot/skills/") && relative.ends_with("/SKILL.md") {
                                Some(("skill", Some("copilot")))
                            } else if (relative.starts_with(".gemini/agents/") || relative.starts_with(".antigravity/agents/")) &&
                                      (relative.ends_with(".md") || relative.ends_with(".agent.md")) {
                                Some(("agent", Some("antigravity")))
                            } else if (relative.starts_with(".gemini/skills/") || relative.starts_with(".antigravity/skills/")) &&
                                      relative.ends_with("/SKILL.md") {
                                Some(("skill", Some("antigravity")))
                            } else if relative.starts_with("skills/") && relative.ends_with("/SKILL.md") {
                                Some(("skill", None))
                            } else if relative.starts_with("agents/") && relative.ends_with(".agent.md") {
                                Some(("agent", None))
                            } else if relative.starts_with(".agents/skills/") && relative.ends_with("/SKILL.md") {
                                Some(("skill", Some("agents")))
                            } else if relative.starts_with(".github/skills/") && relative.ends_with("/SKILL.md") {
                                Some(("skill", Some("copilot")))
                            } else {
                                None
                            }
                        }
                    };

                    if let Some((kind, source_harness)) = matched {
                        if let Ok(meta) = std::fs::metadata(&path) {
                            results.push(serde_json::json!({
                                "id": relative,
                                "name": name,
                                "kind": kind,
                                "path": relative,
                                "sourceHarness": source_harness,
                                "filePath": path.to_string_lossy(),
                                "size": meta.len(),
                                "modifiedAt": meta.modified()
                                    .ok()
                                    .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                                    .map(|d| d.as_millis())
                                    .unwrap_or(0),
                            }));
                        }
                    }
                }
            }
        }
    }
}

/// GET /api/repo-assets/discover?repoPath=...
/// Scan repo for .copilot directory and its contents
async fn discover(
    State(state): State<AppState>,
    Query(query): Query<RepoPathQuery>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let repo_path = query.repo_path.unwrap_or_else(|| {
        state.config.engine_root.to_string_lossy().to_string()
    });
    let root = PathBuf::from(&repo_path);

    if !root.is_dir() {
        return Err(ApiError::NotFound("Repository path not found".to_string()));
    }

    let mut assets = vec![];

    // Scan root files
    for root_file in &["AGENTS.md", "guidelines.md"] {
        let full_path = root.join(root_file);
        if full_path.is_file() {
            if let Ok(meta) = std::fs::metadata(&full_path) {
                assets.push(serde_json::json!({
                    "id": root_file,
                    "name": *root_file,
                    "kind": "config",
                    "path": root_file,
                    "sourceHarness": serde_json::Value::Null,
                    "filePath": full_path.to_string_lossy(),
                    "size": meta.len(),
                    "modifiedAt": meta.modified()
                        .ok()
                        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                        .map(|d| d.as_millis())
                        .unwrap_or(0),
                    "harnesses": HARNESS_NAMES.iter().map(|h| serde_json::json!({
                        "harness": h,
                        "installed": false,
                        "installedAt": null,
                    })).collect::<Vec<_>>(),
                }));
            }
        }
    }

    // Scan recursive directories
    scan_assets_dir(&root, &root, &mut assets, 0);

    // Deduplicate by id
    let mut seen = std::collections::HashSet::new();
    let unique_assets: Vec<serde_json::Value> = assets.into_iter()
        .filter(|a| {
            let id = a["id"].as_str().unwrap_or("").to_string();
            if seen.contains(&id) {
                false
            } else {
                seen.insert(id);
                true
            }
        })
        .collect();

    // Sort by path
    let mut sorted = unique_assets;
    sorted.sort_by(|a, b| {
        a["path"].as_str().unwrap_or("").cmp(b["path"].as_str().unwrap_or(""))
    });

    Ok(Json(serde_json::json!({
        "repoPath": repo_path,
        "assets": sorted,
        "availableHarnesses": HARNESS_NAMES,
        "count": sorted.len(),
    })))
}

/// POST /api/repo-assets/install
/// Read body, copy asset files from catalog to repo
async fn install(
    State(state): State<AppState>,
    Json(body): Json<InstallBody>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let repo_path = body.repo_path.unwrap_or_default();
    let asset_id = body.asset_id.unwrap_or_default();
    let source_path = body.source_path.unwrap_or_default();

    if repo_path.is_empty() {
        return Err(ApiError::BadRequest("repoPath is required".to_string()));
    }
    if asset_id.is_empty() {
        return Err(ApiError::BadRequest("assetId is required".to_string()));
    }

    let root = PathBuf::from(&repo_path);
    if !root.is_dir() {
        return Err(ApiError::NotFound("Repository path not found".to_string()));
    }

    let dest_path = root.join(&asset_id);

    if let Some(parent) = dest_path.parent() {
        std::fs::create_dir_all(parent)
            .map_err(|e| ApiError::Internal(e.into()))?;
    }

    if !source_path.is_empty() {
        let src = PathBuf::from(&source_path);
        if src.is_file() {
            std::fs::copy(&src, &dest_path)
                .map_err(|e| ApiError::Internal(e.into()))?;
        } else if src.is_dir() {
            // Copy directory contents recursively (simple version)
            copy_dir_recursive(&src, &dest_path)?;
        } else {
            // Create empty file as placeholder
            std::fs::write(&dest_path, "")
                .map_err(|e| ApiError::Internal(e.into()))?;
        }
    } else {
        // Use catalog source from engine_root
        let catalog_path = state.config.engine_root.join("catalog-assets").join(&asset_id);
        if catalog_path.is_file() {
            if let Some(parent) = dest_path.parent() {
                std::fs::create_dir_all(parent)
                    .map_err(|e| ApiError::Internal(e.into()))?;
            }
            std::fs::copy(&catalog_path, &dest_path)
                .map_err(|e| ApiError::Internal(e.into()))?;
        } else {
            // Create empty file
            std::fs::write(&dest_path, "")
                .map_err(|e| ApiError::Internal(e.into()))?;
        }
    }

    let meta = std::fs::metadata(&dest_path)
        .map_err(|e| ApiError::Internal(e.into()))?;

    Ok(Json(serde_json::json!({
        "ok": true,
        "assetId": asset_id,
        "harness": body.harness,
        "path": dest_path.to_string_lossy(),
        "size": meta.len(),
    })))
}

fn copy_dir_recursive(src: &Path, dest: &Path) -> Result<(), ApiError> {
    std::fs::create_dir_all(dest).map_err(|e| ApiError::Internal(e.into()))?;
    if let Ok(entries) = std::fs::read_dir(src) {
        for entry in entries.flatten() {
            let entry_path = entry.path();
            let name = entry.file_name();
            let dest_entry = dest.join(&name);
            if let Ok(ftype) = entry.file_type() {
                if ftype.is_dir() {
                    copy_dir_recursive(&entry_path, &dest_entry)?;
                } else if ftype.is_file() {
                    std::fs::copy(&entry_path, &dest_entry)
                        .map_err(|e| ApiError::Internal(e.into()))?;
                }
            }
        }
    }
    Ok(())
}
